jupyterstan/__init__.py

```diff
@@ -1,38 +1,96 @@
+import argparse
 import datetime
-import humanize
+import inspect
+import os
+import pickle
+import re
+from hashlib import sha256
+from pathlib import Path
 from typing import Tuple, Dict
 
-import argparse
-
+import humanize
+import pystan
 from IPython.core.magic import Magics, cell_magic, magics_class
-
 from IPython.utils.capture import capture_output
+import shutil
 
-import pystan
+
+class StanModelCacheClass(object):
+    def __init__(self):
+        self.cache_path = os.path.join(Path.home(), '.cache', 'stan')
+        os.makedirs(self.cache_path, exist_ok=True)
+
+    @classmethod
+    def compile_and_store(cls, model_code, model_name, cache_fn, **kwargs):
+        sm = pystan.StanModel(model_code=model_code, model_name=model_name, **kwargs)
+        with open(cache_fn, 'wb') as f:
+            pickle.dump(sm, f)
+        return sm
+
+    def get_or_create(self, model_code, cache_file_name=None, recompile=False, model_name=None, **kwargs):
+        """Use just as you would `stan`"""
+
+        cache_file_name = cache_file_name or 'model-{digest}.pkl'
+        normalized_code = re.sub(r'\s+', ' ', model_code).strip().encode('ascii')
+        hasher = sha256(normalized_code)
+        hasher.update(repr(kwargs).encode('ascii'))
+        digest = hasher.hexdigest()
+
+        cache_fn = cache_file_name.format(digest=digest, model_name=kwargs.get('model_name'))
+        cache_fn = os.path.join(self.cache_path, cache_fn)
+
+        if recompile:
+            sm = self.compile_and_store(model_code, model_name, cache_fn, **kwargs)
+            created = True
+        else:
+            try:
+                sm = pickle.load(open(cache_fn, 'rb'))
+            except Exception as e:
+                if not isinstance(e, FileNotFoundError):
+                    pystan.logger.warning("Problems loading cached model. Recompiling.")
+                sm = self.compile_and_store(model_code, model_name, cache_fn, **kwargs)
+                created = True
+            else:
+                pystan.logger.debug("Using cached StanModel '{}'".format(cache_fn))
+                created = False
+
+        return sm, created
+
+    def clean(self):
```

Comment:
Thanks for adding this clean method to the class, but at the moment there's no real way of invoking this (since most people won't be importing anything and use it as an extension only).

Could you add a line magic like `%stan_clear_cache` or something along those lines?

Otherwise, I can also merge this straight away and implement that later.